### ISSUE TITLE
DataParallel Support for DallE (RuntimeError: Unconvertible NCCL type)

### ISSUE THREAD

**Issue by johnmccain** [lucidrains/DALLE-pytorch#36]
Hello! I was trying to train DallE with DataParallel on a multi GPU machine but encountered the following error:

```
Traceback (most recent call last):
  File "minimal_reproduction.py", line 67, in <module>
    main()
  File "minimal_reproduction.py", line 62, in main
    loss = dalle(caption_ids, image_tokens, mask = mask, return_loss = True)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py", line 160, in forward
    replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/nn/parallel/data_parallel.py", line 165, in replicate
    return replicate(module, device_ids, not torch.is_grad_enabled())
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/nn/parallel/replicate.py", line 103, in replicate
    buffer_copies_not_rg = _broadcast_coalesced_reshape(buffers_not_rg, devices, detach=True)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/nn/parallel/replicate.py", line 67, in _broadcast_coalesced_reshape
    return comm.broadcast_coalesced(tensors, devices)
  File "/home/ubuntu/.local/lib/python3.8/site-packages/torch/nn/parallel/comm.py", line 56, in broadcast_coalesced
    return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
RuntimeError: Unconvertible NCCL type
```

Does this implementation of DallE support multi GPU training with DataParallel? Or is the error between the keyboard and the chair 😄? I tested this with `dalle-pytorch==0.0.59`.
I didn't encounter this on my local machine, but that only has a single GPU so not exactly apples-to-apples.

---
Here's a gist with a training script to reproduce the error: https://gist.github.com/johnmccain/ae2d9f376abff90dfd5cd4c9b36fa7e5

Environment (from https://github.com/pytorch/pytorch/blob/master/torch/utils/collect_env.py):
```
PyTorch version: 1.7.1+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration:
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 450.80.02
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] dalle-pytorch==0.0.59
[pip3] numpy==1.17.4
[pip3] torch==1.7.1+cu110
[pip3] torchaudio==0.7.2
[pip3] torchvision==0.8.2+cu110
[conda] Could not collect
```

**Comment by lucidrains** [comment]
@johnmccain i'm going to get a training framework in place soon! been distracted with all the great papers streaming out

**Comment by kobiso** [comment]
@johnmccain I've experienced the same error when I used `torch.nn.DataParallel`.
The error was hard to debug because it did not point out the exact error location.

Instead, I decided to use [horovod](https://github.com/horovod/horovod) so that I can extend it from multi-gpu to multi-server training.
It works fine for multi-gpu training till now (haven't tried multi-server training yet).

@lucidrains I know you will work on multi-gpu training script, you can consider horovod as an option :)
Horovod training seems very fast and easy to debug.

### One thing I had to fix to use horovod
- I've got below error in here 
https://github.com/lucidrains/DALLE-pytorch/blob/f03396334229111547a7dae2bd3561490325897d/dalle_pytorch/dalle_pytorch.py#L464-L467
- I guess horovod does not support `torch.cuda.BoolTensor` for `broadcast_parameters()`.

```
[0]<stderr>:Traceback (most recent call last):
[0]<stderr>:  File "train_DALLE.py", line 345, in <module>
[0]<stderr>:    main()
[0]<stderr>:  File "train_DALLE.py", line 212, in main
[0]<stderr>:    hvd.broadcast_parameters(dalle.state_dict(), root_rank=0)
[0]<stderr>:  File "/opt/conda/lib/python3.7/site-packages/horovod/torch/functions.py", line 53, in broadcast_parameters
[0]<stderr>:    handle = broadcast_async_(p, root_rank, name)
[0]<stderr>:  File "/opt/conda/lib/python3.7/site-packages/horovod/torch/mpi_ops.py", line 693, in broadcast_async_
[0]<stderr>:    return _broadcast_async(tensor, tensor, root_rank, name)
[0]<stderr>:  File "/opt/conda/lib/python3.7/site-packages/horovod/torch/mpi_ops.py", line 598, in _broadcast_async
[0]<stderr>:    function = _check_function(_broadcast_function_factory, tensor)
[0]<stderr>:  File "/opt/conda/lib/python3.7/site-packages/horovod/torch/mpi_ops.py", line 83, in _check_function
[0]<stderr>:    raise ValueError('Tensor type %s is not supported.' % tensor.type())
[0]<stderr>:ValueError: Tensor type torch.cuda.BoolTensor is not supported.
[1]<stderr>:Traceback (most recent call last):
[1]<stderr>:  File "/opt/conda/lib/python3.7/site-packages/horovod/torch/mpi_ops.py", line 601, in _broadcast_async
[1]<stderr>:    tensor, output, root_rank, name.encode() if name is not None else _NULL)
[1]<stderr>:RuntimeError: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to al
lreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exceptio
n, you should see the exception in the log before the first shutdown message.
```

- I solved this problem by simply changing the `torch.cuda.BoolTensor` to `torch.uint8` type.
```
logits_mask = (
            ((seq_range >= text_seq_len) & (logits_range < num_text_tokens)) |
            ((seq_range < text_seq_len) & (logits_range >= num_text_tokens))
        ).type(torch.uint8)
```

**Comment by johnmccain** [comment]
Thanks for the tip @kobiso! I'll have to give horovod a try for this model. I didn't have any success with DistributedDataParallel either--I wonder if it's caused by the same issue (the error message I got with DDP was even more cryptic than with DataParallel).
After a bit of further digging, I suspect the error is related to this PyTorch issue: https://github.com/pytorch/pytorch/issues/46745
Looks like a fix was applied but after the 1.7.1 release. I'll give this a try with the latest nightly PyTorch build sometime soon and see if that resolves the issue.

**Comment by johnmccain** [comment]
With the release of PyTorch 1.8 today improving the reliability of NCCL for distributed training, I'm closing this issue.
